Improve doc for det and add test for legacy

sylee957 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->

#12854 #12363
Closes #15263 

#### Brief description of what is fixed or changed

~~The deprecation for the keys was reverted in #12854, but I think that it should be reintroduced.~~

https://github.com/sympy/sympy/blob/4422baac2db5f51cffae5b11e1efb8307377bbc1/sympy/matrices/matrices.py#L398-L405

These lines had just make it more tolerant for typos, but I don't think they should stay forever.
Encouraging the use of only one keyword will make other developers easy to track down the usages.

I have also found that the typo 'bareis' in docstring had been reintroduced after the merge.

#### Other comments

~~I have added stub issue number, because it does not allow to build without entering issue number.
I will change it to proper number after getting a new PR ticket.~~

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

sylee957 committed:
Fix Typo in docstring

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -390,16 +390,36 @@ def det(self, method="bareiss", iszerofunc=None):
+                feature="det(method='{}')".format(method_warn),
+                useinstead="det(method='{}')".format(method),
+                deprecated_since_version="1.4",
+                issue=15256
```
This should be the number of a deprecation removal issue (see https://github.com/sympy/sympy/wiki/Deprecating-policy).

moorepants commented:
I'm generally -1 for deprecations like this. It only really serves to break old code and annoy users. Why not just change the documentation to seem like we only accept lower case keywords yet accept uppercase anyways? Then best of both worlds, no breaking code, no deprecation, and people are encouraged to do a better practice. Same thing with the misspelling. Just quietly accept the misspelled word if that is what we've always done, correct the docs.

sylee957 commented:
@moorepants 
If it is not going to deprecated, then I think deleted test cases should be readded to ensure minimal support, besides the docs

moorepants commented:
Yep, test cases should still be in place for all functionality even if we don't advertise it.

sylee957 committed:
Better Documentation

sylee957 committed:
Write docstrings

sylee957 committed:
Added a line to fix sphinx bullet list fail

sylee957 committed:
Moved to note directive

sylee957 committed:
Added Returns section

sylee957 committed:
Added Raises Section

sylee957 committed:
optional for optional key

sylee957 committed:
Misc changes to method doc

sylee957 committed:
Added iszerofunc section

sylee957 committed:
Added additional indent to the bullet lists

sylee957 commented:
@moorepants 
I have moved the deprecated keys to Notes, but I think `Notes` section is rendered as a plane paragraph in sympy, so I had added to `.. note::` directive instead.

@jksuom 
I have also added iszerofunc section for changes made in #15207. And other Exceptions and Returns sections to make a complete document.

However, from this moment, docs are not getting rendered because of the issue #15272,
so I have partially merged 90fe2a6fce48379ce1b5457820de87153eed1160 for getting doc render for this work, and got the result as below

![image](https://user-images.githubusercontent.com/34944973/45875722-772d3280-bdd3-11e8-8666-1306d8cfc5a0.png)

Any suggestions?

sylee957 committed:
Added test cases for legacy support for det().

jksuom commented:
This is not very helpful:

    ! LaTeX Error: Too deeply nested.

but it looks like `.. autoclass :: MatrixDeterminant` is causing it. Maybe it would work without the `:members:` part.

sylee957 committed:
Removed bullet list to fix too-deeply-nested latex error

sylee957 commented:
@jksuom 

I think the error can be fixed by flattening the docs.
In this way, it will work fine with `autoclass`.

![image](https://user-images.githubusercontent.com/34944973/46077763-f86a3800-c1cc-11e8-858b-827cccb6c074.png)

jksuom commented:
This looks ok to me though I am not sure of the language. Perhaps someone else could also take a look.

sidhantnagpal commented on sympy/matrices/matrices.py:
```diff
@@ -384,15 +384,56 @@ def cofactor_matrix(self, method="berkowitz"):
 
+        Parameters
+        ==========
+        method : string, optional
```
A new line should be added before `method` (and similarly for `Returns` and `Raises`).

sidhantnagpal commented on sympy/matrices/matrices.py:
```diff
@@ -384,15 +384,56 @@ def cofactor_matrix(self, method="berkowitz"):
+        Parameters
+        ==========
+        method : string, optional
+            Specifies an algorithm to use for computing matrix determinant.
```
`Specifies the algorithm used for computing the matrix determinant.`

sidhantnagpal commented on sympy/matrices/matrices.py:
```diff
@@ -384,15 +384,56 @@ def cofactor_matrix(self, method="berkowitz"):
+            Specifies an algorithm to use for computing matrix determinant.
+
+            If the matrix is at most 3x3, the key is ignored and a hard-coded
+            formula is used. Otherwise, it will be defaulted with 'bareiss'.
```
Possibly: `If the matrix is at most 3x3, a hard-coded formula is used and the specified method is ignored. Otherwise, it defaults to 'bareiss'.`

sidhantnagpal commented on sympy/matrices/matrices.py:
```diff
@@ -384,15 +384,56 @@ def cofactor_matrix(self, method="berkowitz"):
+            formula is used. Otherwise, it will be defaulted with 'bareiss'.
+
+            If it is set with 'bareiss', Bareiss' fraction-free algorithm would
+            be used
```
Ending with a period (`.`) and probably better worded as:
`If it is set to 'bareiss', Bareiss' fraction-free algorithm will be used.` (similary for 'lu' and 'berkowitz').

sidhantnagpal commented on sympy/matrices/matrices.py:
```diff
@@ -384,15 +384,56 @@ def cofactor_matrix(self, method="berkowitz"):
+
+        iszerofunc : FunctionType or None, optional
+            Specifies zero testing method used for 'bareiss' or 'lu'.
+            If unspecified, it will be defaulted into None
```
This line might not be necessary as it is self-documented in the method signature.

sidhantnagpal commented on sympy/matrices/matrices.py:
```diff
@@ -384,15 +384,56 @@ def cofactor_matrix(self, method="berkowitz"):
+            Specifies zero testing method used for 'bareiss' or 'lu'.
+            If unspecified, it will be defaulted into None
+
+            If it is set to None, it will be defaulted to _iszero when when
```
"...if the method is set to..." instead of "...when when...".

sidhantnagpal commented on sympy/matrices/matrices.py:
```diff
@@ -384,15 +384,56 @@ def cofactor_matrix(self, method="berkowitz"):
+
+        Raises
+        ======
+        ValueError :
```
"ValueError : if unrecognized keys are given for \`method\` or \`iszerofunc\`" (similarly for `NonSquareMatrixError`).

sylee957 committed:
Added extra spacing after paragraphs

sylee957 committed:
Changed phrase

sylee957 committed:
Change phrase

sylee957 committed:
Changed phrase

sylee957 committed:
Deleted unnecessary words

sylee957 committed:
Corrected typo

sylee957 committed:
Edited the raises section

sylee957 committed:
Use inline literals

sidhantnagpal commented on sympy/matrices/matrices.py:
```diff
@@ -384,15 +384,55 @@ def cofactor_matrix(self, method="berkowitz"):
+        ======
+
+        ValueError
+            If unrecognized keys are given for ``method`` or ``iszerofunc``
```
Ending with a period.

sidhantnagpal commented on sympy/matrices/matrices.py:
```diff
@@ -384,15 +384,55 @@ def cofactor_matrix(self, method="berkowitz"):
+
+        iszerofunc : FunctionType or None, optional
+            If it is set to ``None``, it will be defaulted to ``_iszero`` if the
+            method is set to ``bareiss``, and ``_is_zero_after_expand_mul`` if
```
The usage of keys could possibly pointed out as strings. ("bareiss" and similar)

sidhantnagpal commented on sympy/matrices/matrices.py:
```diff
@@ -384,15 +384,55 @@ def cofactor_matrix(self, method="berkowitz"):
+
+            .. note::
+                For backward compatibility, legacy keys like **bareis** ,
+                **det_lu** can still be used to indicate corresponding
```
I don't think these keys should be bold as their use is not encouraged. (..."bareis" and "det_lu"...)

sidhantnagpal commented on sympy/matrices/matrices.py:
```diff
@@ -384,15 +384,55 @@ def cofactor_matrix(self, method="berkowitz"):
+            If it is set to ``bareiss``, Bareiss' fraction-free algorithm will
+            be used.
+
+            If it is set to ``berkowitz``, Berkowitz' algorithm would be used.
```
...will be used. (similarly for "lu")

sylee957 committed:
Unbold some phrases

sylee957 committed:
Add apostrophes for keys in string form

sylee957 committed:
Add apostrophes 2

sylee957 committed:
Change would to will

sylee957 commented:
![image](https://user-images.githubusercontent.com/34944973/46346503-319a2080-c683-11e8-844b-bdefa69b4225.png)

sidhantnagpal commented on sympy/matrices/matrices.py:
```diff
@@ -384,15 +384,56 @@ def cofactor_matrix(self, method="berkowitz"):
+
+            .. note::
+                For backward compatibility, legacy keys like "bareis" ,
+                "det_lu" can still be used to indicate corresponding
```
..."bareis" and "det_lu"...indicate the corresponding...

sidhantnagpal commented on sympy/matrices/matrices.py:
```diff
@@ -384,15 +384,56 @@ def cofactor_matrix(self, method="berkowitz"):
+                For backward compatibility, legacy keys like "bareis" ,
+                "det_lu" can still be used to indicate corresponding
+                methods.
+                And the keys are also case-insensitive for now.
```
I am not sure if it is necessary to mention this in the documentation (even if it is handled in code).

sylee957 replied:
Well, I originally had thought that the intent to make keys case-insensitive was not a good idea, as it had added some redundant code.
And also some people would use case-sensitive searches to track down their usage of keywords, expecting that every function accepts only the exact keys.

sidhantnagpal replied:
Maybe @moorepants would like to comment on this.

It seems that your last commit introduced trailing whitespaces.

sidhantnagpal commented on sympy/matrices/matrices.py:
```diff
@@ -384,15 +384,56 @@ def cofactor_matrix(self, method="berkowitz"):
+                "det_lu" can still be used to indicate corresponding
+                methods.
+                And the keys are also case-insensitive for now.
+                However, we would discourage you to use the keys in the manner.
```
Possibly reword this in an optimistic way `However, it is suggested to use the proper keys for specifying the method.` and write it in continuity inside the note (without starting a new paragraph).

sidhantnagpal reviewed (commented):
These are my final comments. Let's wait to see if anyone would like to comment.

sylee957 committed:
Changed the phrase in note

sylee957 committed:
Add inline literal

sylee957 committed:
Changed comma to add

sylee957 committed:
proper to precise?

sylee957 committed:
Added 'the'

sylee957 committed:
Remove trailing whitespace

jksuom commented:
+1
